internal/controller/isbservicerollout_controller.go

```diff
@@ -177,28 +191,186 @@ func (r *ISBServiceRolloutReconciler) reconcile(ctx context.Context, isbServiceR
 		Spec: isbServiceRollout.Spec.InterStepBufferService.Spec,
 	}
 
-	err := kubernetes.ApplyCRSpec(ctx, r.restConfig, &obj, "interstepbufferservices")
+	existingISBServiceDef, err := kubernetes.GetCR(ctx, r.restConfig, &newISBSVCDef, "interstepbufferservices")
 	if err != nil {
-		numaLogger.Errorf(err, "failed to apply CR: %v", err)
-		return err
+		// create object as it doesn't exist
+		if apierrors.IsNotFound(err) {
+			numaLogger.Debugf("ISBService %s/%s doesn't exist so creating", isbServiceRollout.Namespace, isbServiceRollout.Name)
+			isbServiceRollout.Status.MarkPending()
+
+			err = kubernetes.CreateCR(ctx, r.restConfig, &newISBSVCDef, "interstepbufferservices")
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			isbServiceRollout.Status.MarkDeployed(isbServiceRollout.Generation)
+
+		} else {
+			return ctrl.Result{}, fmt.Errorf("error getting ISBService: %v", err)
+		}
+
+	} else {
+		// Object already exists
+		// perform logic related to updating
+
+		newISBSVCDef = *mergeISBService(existingISBServiceDef, &newISBSVCDef)
+
+		// update our Status with the ISBService's Status
+		r.processISBServiceStatus(ctx, existingISBServiceDef, isbServiceRollout)
+
+		//todo: move these to initialization since we don't want to support changing them while running
+		globalConfig, err := config.GetConfigManagerInstance().GetConfig()
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error getting global config: %w", err)
+		}
+		// if I need to update or am in the middle of an update of the ISBService, then I need to make sure all the Pipelines are pausing
+		isbServiceNeedsUpdating, isbServiceIsUpdating, err := r.isISBServiceUpdating(ctx, isbServiceRollout, existingISBServiceDef)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		numaLogger.Debugf("isbServiceNeedsUpdating=%t, isbServiceIsUpdating=%t", isbServiceNeedsUpdating, isbServiceIsUpdating)
+
+		// set the Status appropriately to "Pending" or "Deployed"
+		// if isbServiceNeedsUpdating - this means there's a mismatch between the desired ISBService spec and actual ISBService spec
+		// if there's a generation mismatch - this means we haven't even observed the current generation
+		// we may match the first case and not the second when we've observed the generation change but we're pausing pipelines
+		// we may match the second case and not the first if we need to update something other than ISBService spec
+		if isbServiceNeedsUpdating || isbServiceRollout.Status.ObservedGeneration < isbServiceRollout.Generation {
+			isbServiceRollout.Status.MarkPending()
+		} else {
+			isbServiceRollout.Status.MarkDeployed(isbServiceRollout.Generation)
+		}
```

Comment:
`Pending` now means we haven't updated the spec (or anything else) as we should: either we haven't even observed the latest generation (e.g. maybe we errored out while trying to process it) or we have observed it but simply are waiting for pipelines to pause